### PR TITLE
Add custom matchers for completions (fuzzy search for presentation compiler)

### PR DESCRIPTION
**Pull request by rochala**
Fixes https://github.com/scalameta/metals/issues/4656
Fixes https://github.com/lampepfl/dotty/issues/5507
Fixes https://github.com/lampepfl/dotty/issues/17706

To ensure CI output will be the same on all runners, I'm scheduling runs on all jvm versions + windows.

[test_windows_full]
[test_java8]
[test_java11]
[test_java15]
[test_java17]
[test_java18]
[test_java19]

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 1]
```diff
@@ -503,10 +505,10 @@ object Completion:
       // There are four possible ways for an extension method to be applicable
 
       // 1. The extension method is visible under a simple name, by being defined or inherited or imported in a scope enclosing the reference.
-      val termCompleter = new Completer(Mode.Term, prefix, pos)
-      val extMethodsInScope = termCompleter.scopeCompletions.toList.flatMap:
-        case (name, denots) => denots.collect:
-          case d: SymDenotation if d.isTerm && d.termRef.symbol.is(Extension) => (d.termRef, name.asTermName)
+      val termCompleter = new Completer(Mode.Term, prefix, pos, _.startsWith(prefix))
```
Comment (filipwiech) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:508]: Sorry if the question doesn't make much sense, I'm not really familiar with this code, but does it mean that extension methods will be excluded from the fuzzy search? :slightly_smiling_face: 

In any case, great to see this improvement finally coming to Scala 3 completions! :+1:

Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:508]: This is my oversight. It should also be fuzzy. Great catch 👍🏻

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 2]
```diff
@@ -71,10 +71,11 @@ object Completion:
     mode: Mode,
     rawPrefix: String,
     tpdPath: List[tpd.Tree],
-    untpdPath: List[untpd.Tree]
+    untpdPath: List[untpd.Tree],
+    prefixFilter: Option[Name => Boolean] = None
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:75]:
```suggestion
    customMatcher: Option[Name => Boolean] = None
```
maybe?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 3]
```diff
@@ -317,7 +319,7 @@ object Completion:
    *  For the results of all `xyzCompletions` methods term names and type names are always treated as different keys in the same map
    *  and they never conflict with each other.
    */
-  class Completer(val mode: Mode, val prefix: String, pos: SourcePosition):
+  class Completer(val mode: Mode, val prefix: String, pos: SourcePosition, prefixFilter: Name => Boolean):
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:322]:
```suggestion
  class Completer(val mode: Mode, pos: SourcePosition, matches: Name => Boolean):
```
Do you even need the prefix anymore? Also update scaladoc.

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 4]
```diff
@@ -556,7 +558,9 @@ object Completion:
      *   2. satisfy [[Completion.isValidCompletionSymbol]]
      */
     private def include(denot: SingleDenotation, nameInScope: Name)(using Context): Boolean =
-      nameInScope.startsWith(prefix) &&
+      val sym = denot.symbol
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:561]: unused?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 5]
```diff
@@ -524,10 +526,10 @@ object Completion:
       // There are four possible ways for an extension method to be applicable
 
       // 1. The extension method is visible under a simple name, by being defined or inherited or imported in a scope enclosing the reference.
-      val termCompleter = new Completer(Mode.Term, prefix, pos)
-      val extMethodsInScope = termCompleter.scopeCompletions.toList.flatMap:
-        case (name, denots) => denots.collect:
-          case d: SymDenotation if d.isTerm && d.termRef.symbol.is(Extension) => (d.termRef, name.asTermName)
+      val termCompleter = new Completer(Mode.Term, prefix, pos, prefixFilter)
+      val extMethodsInScope = termCompleter.scopeCompletions.toList.flatMap {
+        case (name, denots) => denots.collect { case d: SymDenotation if d.isTerm => (d.termRef, name.asTermName) }
```
Comment (kasiaMarek) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala:531]: why is that `d.termRef.symbol.is(Extension)` condition part not needed anymore?

Comment (rochala) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala:531]: It is filtered in later parts, but I'll restore it just to align it with the variable

**Review by kasiaMarek — APPROVED** [review]

**Pull request merged**